src/bin/encode.rs

```diff
@@ -0,0 +1,152 @@
+        let mut out = std::io::BufWriter::new(video_file);
+
+        for frame in 0..FRAME_COUNT {
+            write_frame(
+                &mut out,
+                &mut image,
+                &data,
+                frame,
+                Color([255, 100, 16]),
+                Color([64, 128, 255]),
+                Color([255, 255, 255]),
+            );
+        }
+    }
+
+    // Use FFmpeg to convert the raw content into a video
+    let fps = format!("{}", FPS);
+    let hz = format!("{}", HZ);
+    std::process::Command::new("ffmpeg")
+        .args(&["-r", &fps, "-f", "image2pipe", "-i", "video"])
+        .args(&["-f", "u8", "-ar", &hz, "-ac", "1", "-i", "audio"])
+        .args(&["-pix_fmt", "yuv420p", "-y", "-s", "640x360", "out.mp4"])
+        .spawn()
+        .unwrap()
+        .wait()
+        .unwrap();
+
+    std::fs::remove_file("audio").unwrap();
+    std::fs::remove_file("video").unwrap();
+}
+
+fn write_ppm<W: Write>(out: &mut W, buf: &[Color], width: usize, height: usize) {
+    write!(out, "P6\n{} {}\n255\n", width, height).unwrap();
+    for pix in buf {
+        out.write(&pix.0[..]).unwrap();
+    }
+}
+
+fn write_frame<W: Write>(
+    out: &mut W,
+    image: &mut [Color],
+    data: &[u8],
+    frame: usize,
+    bg_color: Color,
+    wave_color: Color,
+    scan_color: Color,
+) {
+    // Compute the offsets
+    let (col, off) = if frame < FRAME_COUNT / 4 {
+        (WIDTH * frame * 2 / FRAME_COUNT, 0)
+    } else if frame < 3 * FRAME_COUNT / 4 {
+        (
+            WIDTH / 2,
+            (frame - FRAME_COUNT / 4) * WIDTH * 2 / FRAME_COUNT,
+        )
+    } else {
+        (WIDTH * (frame - FRAME_COUNT / 2) * 2 / FRAME_COUNT, WIDTH)
+    };
+
+    // Draw the background
+    for i in 0..WIDTH * HEIGHT {
+        let idx = (i % HEIGHT) * WIDTH + (i / HEIGHT);
+        let x = data[i + off * HEIGHT];
+        image[idx] = bg_color * x;
+    }
+
+    // Draw the cursor
+    for row in 0..HEIGHT {
+        image[row * WIDTH + col] = scan_color;
+    }
+
+    // Draw the waveform
+    let x = WIDTH * frame * 2 / FRAME_COUNT;
+    for row in 0..HEIGHT - 1 {
+        let x0 = data[x * HEIGHT + row];
+        let x1 = data[x * HEIGHT + row + 1];
+        let col = row * WIDTH / HEIGHT;
+        let (bot, top) = (x0.min(x1) as usize, x0.max(x1) as usize);
+        let mid = (bot + top) / 2;
+        for r in bot..mid + 1 {
+            image[(255 - r) * WIDTH + col] = wave_color;
+        }
+        for r in mid..top + 1 {
+            image[(255 - r) * WIDTH + col + 1] = wave_color;
+        }
+    }
+
+    // Output an image
+    write_ppm(out, &image, WIDTH, HEIGHT);
+}
+
+impl std::ops::Mul<u8> for Color {
+    type Output = Color;
+    fn mul(self, rhs: u8) -> Color {
+        let Color(lhs) = self;
+        let r = lhs[0] as u16 * rhs as u16;
+        let g = lhs[1] as u16 * rhs as u16;
+        let b = lhs[2] as u16 * rhs as u16;
+        Color([(r >> 8) as u8, (g >> 8) as u8, (b >> 8) as u8])
```

Comment:
While it isn't a good idea to do this, leaving off the `>> 8` makes for some really pretty color schemes, I might keep this in mind as a just for fun thing later on.

Reply:
Heh, that's not fixed 0.8 multiplication, but yeah.